### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by nblach**

### CODE REVIEW

File: src/Schedgen2/mpi_colls.py [hunk 1]
```diff
@@ -56,103 +27,185 @@ def dissemination(comm_size, datasize, tag):
...
-            datasize // comm_size
-            if datasize % comm_size == 0
-            else datasize // comm_size + 1
+
+def reduce(
+    comm_size: int,
+    datasize: int,
+    tag: int = 42,
+    ptrn: str = "binomialtree",
+    unbalanced: bool = False,
```
Comment (tim0s) [inline src/Schedgen2/mpi_colls.py]: what is this unbalanced flag?

File: src/Schedgen2/mpi_colls.py [hunk 2]
```diff
@@ -56,103 +27,185 @@ def dissemination(comm_size, datasize, tag):
...
-            sources,
-            destinations,
-            data_sizes_receive,
-            data_sizes_send,
-            base_tag,
-            last_dependency=dependency,
-            compute_time_dependency=ctd,
-        )
 
+def outcast(
```
Comment (tim0s) [inline src/Schedgen2/mpi_colls.py]: this should not be here, there is no MPI_Outcast function in the MPI standard, so there should be no outcast function in this file. please remove.

File: src/Schedgen2/mpi_colls.py [hunk 3]
```diff
@@ -56,103 +27,185 @@ def dissemination(comm_size, datasize, tag):
...
-            comm,
-            rank,
-            sources,
-            destinations,
-            data_sizes_receive,
-            data_sizes_send,
-            base_tag,
-            compute_time_dependency=ctd,
-        )
+def incast(
```
Comment (tim0s) [inline src/Schedgen2/mpi_colls.py]: same as below, MPI_Incast does not exist, remove

File: src/Schedgen2/mpi_colls.py [hunk 4]
```diff
@@ -56,103 +27,185 @@ def dissemination(comm_size, datasize, tag):
...
+    else:
+        raise ValueError(f"reduce with pattern {ptrn} not implemented")
+
+
+def bcast(
+    comm_size: int,
+    datasize: int,
+    tag: int = 42,
+    ptrn: str = "binomialtree",
+    unbalanced: bool = False,
```
Comment (tim0s) [inline src/Schedgen2/mpi_colls.py]: MPI_Bcast has no argument for "unbalanced" anything, what is this supposed to do?

Comment (nblach) [reply src/Schedgen2/mpi_colls.py]: I will rename the unbalanced to randomize_data, to make it more intuitive. Some benchmarks want to randomise the data sent to each process, which is enabled using that flag. This is not part of the MPI_Bcast, but I don't see an effective way to map this 1-to-1

File: src/Schedgen2/mpi_colls.py (hunk 1 above)
Comment (nblach) [reply src/Schedgen2/mpi_colls.py]: Please see my response to the bcast comment

File: src/Schedgen2/mpi_colls.py (hunk 2 above)
Comment (nblach) [reply src/Schedgen2/mpi_colls.py]: will move to a new file

File: src/Schedgen2/mpi_colls.py (hunk 3 above)
Comment (nblach) [reply src/Schedgen2/mpi_colls.py]: will move to a new file

**Pull request merged**